tests/Unit/Evolution/Systems/ScalarWave/Test_Constraints.cpp

```diff
@@ -0,0 +1,344 @@
+               Spectral::Quadrature::GaussLobatto};
+
+  using Affine = domain::CoordinateMaps::Affine;
+  using Affine3D =
+      domain::CoordinateMaps::ProductOf3Maps<Affine, Affine, Affine>;
+  const auto coord_map =
+      domain::make_coordinate_map<Frame::Logical, Frame::Inertial>(Affine3D{
+          Affine{-1.0, 1.0, lower_bound[0], upper_bound[0]},
+          Affine{-1.0, 1.0, lower_bound[1], upper_bound[1]},
+          Affine{-1.0, 1.0, lower_bound[2], upper_bound[2]},
+      });
+
+  // Set up coordinates
+  const auto x_logical = logical_coordinates(mesh);
+  const auto x = coord_map(x_logical);
+  // Arbitrary time for time-independent solution.
+  const double t = 0.;
+
+  // Evaluate analytic solution
+  const auto vars = solution.variables(
+      x, t,
+      tmpl::list<ScalarWave::Pi, ScalarWave::Phi<dim>, ScalarWave::Psi>{});
+
+  // Compute derivatives d_psi, d_phi, d_pi, and d_gauge_function numerically
+  Variables<VariablesTags> sw_vars(data_size);
+  get<ScalarWave::Psi>(sw_vars) = get<ScalarWave::Psi>(vars);
+  get<ScalarWave::Pi>(sw_vars) = get<ScalarWave::Pi>(vars);
+  get<ScalarWave::Phi<dim>>(sw_vars) = get<ScalarWave::Phi<dim>>(vars);
+
+  // Compute numerical derivatives of psi,pi,phi,H
+  const auto sw_derivs =
+      partial_derivatives<VariablesTags, VariablesTags, dim, Frame::Inertial>(
+          sw_vars, mesh, coord_map.inv_jacobian(x_logical));
+  const auto& d_phi = get<
+      Tags::deriv<ScalarWave::Phi<dim>, tmpl::size_t<dim>, Frame::Inertial>>(
+      sw_derivs);
+
+  // Get the constraint, and check that it vanishes to error_tolerance
+  auto two_index_constraint =
+      make_with_value<tnsr::ij<DataVector, dim, Frame::Inertial>>(
+          x, std::numeric_limits<double>::signaling_NaN());
+  ScalarWave::two_index_constraint(make_not_null(&two_index_constraint), d_phi);
+
+  Approx numerical_approx =
+      Approx::custom().epsilon(error_tolerance).scale(1.0);
+  CHECK_ITERABLE_CUSTOM_APPROX(
+      two_index_constraint,
+      make_with_value<decltype(two_index_constraint)>(x, 0.0),
+      numerical_approx);
+}
+
+// Test the return-by-value four-index constraint function using random values
+template <size_t Dim>
+void test_one_index_constraint_random(
+    const DataVector& used_for_size) noexcept {
+  pypp::check_with_random_values<1>(
+      static_cast<tnsr::i<DataVector, Dim, Frame::Inertial> (*)(
+          const tnsr::i<DataVector, Dim, Frame::Inertial>&,
+          const tnsr::i<DataVector, Dim, Frame::Inertial>&)>(
+          &ScalarWave::one_index_constraint<Dim>),
+      "TestFunctions", "one_index_constraint", {{{-10.0, 10.0}}},
+      used_for_size);
+}
+
+// Test the return-by-reference four-index constraint
+// by comparing to a time-independent analytic solution (e.g. Kerr-Schild)
+template <typename Solution>
+void test_one_index_constraint_analytic(
+    const Solution& solution, const size_t grid_size_each_dimension,
```

Comment:
Can you just test both constraints and their compute items in the same function to reduce code duplication?

Reply:
merged both functions